doc/concepts/locks.rst

```diff
@@ -0,0 +1,49 @@
+Locks
+=====
+
+Given two rules that are independent, Dune will assume that their
+associated actions can be run concurrently. Two rules are considered
+independent if neither of them depend on the other, either directly or
+through a chain of dependencies. This basic assumption allows Dune to
+parallelize the build.
+
+However, it is sometimes the case that two independent rules cannot be
+executed concurrently. For instance this can happen for more
+complicated tests. In order to prevent Dune from running the
+actions at the same time, you can specify that both actions take the
+same lock:
+
+.. code:: dune
+
+    (rule
+     (alias  runtest)
+     (deps   foo)
+     (locks  m)
+     (action (run test.exe %{deps})))
+
+    (alias
+     (rule   runtest)
+     (deps   bar)
+     (locks  m)
+     (action (run test.exe %{deps})))
+
+Dune will make sure that the executions of ``test.exe foo`` and
+``test.exe bar`` are serialized.
+
+Although they don't live in the filesystem, lock names are interpreted as file
+names. So for instance ``(with-lock m ...)`` in ``src/dune`` and ``(with-lock
+../src/m)`` in ``test/dune`` refer to the same lock.
+
+Note also that locks are per build context. So if your workspace has two build
+contexts setup, the same rule might still be executed concurrently between the
```

Comment:
```suggestion
context setups, the same rule might still be executed concurrently between the
```

Reply:
to clarify: there's a notion of "build context", and a workspace has two of them set up. how can we word this?

Reply:
Hmmmm… Perhaps: “…if your workspace has two build-context setups…” or “…if you have set up to build contexts in your workspace” ??